apollo-federation/src/connectors/json_selection/parser.rs

```diff
@@ -707,12 +712,34 @@ impl PathList {
             ));
         }
 
-        // Universal optional operator: ? (note: we parse this before other operators to avoid conflicts)
-        if let Ok((suffix, question)) = ranged_span("?")(input) {
-            // Parse whatever comes after the ? normally
-            let (remainder, tail) = Self::parse_with_depth(suffix, depth)?;
-            let full_range = merge_ranges(question.range(), tail.range());
-            return Ok((remainder, WithRange::new(Self::Question(tail), full_range)));
+        // At any depth, if the next token is ? but not the PathList::Question
+        // kind, we terminate path parsing so the hypothetical ?? or ?! tokens
+        // have a chance to be parsed as infix operators.
+        if input.fragment().starts_with("??") || input.fragment().starts_with("?!") {
+            return Ok((input, WithRange::new(Self::Empty, range_if_empty)));
+        }
+
+        if let Ok((suffix, question)) = ranged_span("?")(input.clone()) {
+            let (remainder, rest) = Self::parse_with_depth(suffix, depth + 1)?;
```

Comment:
Hmm, I'm now realizing/remembering these parser changes won't cover `NamedFieldSelection`s like `age?` in `author { name age? }`, so we really do need the `NamedSelection` unification from #7945 to do `?` syntax right.